executor/common_linux.h

```diff
@@ -1570,6 +1570,293 @@ static long syz_init_net_socket(volatile long domain, volatile long type, volati
+		return;
+	}
+	}
+
+	char status[0xf9] = {0};
+	hci_send_event_cmd_complete(fd, hdr->opcode, status, sizeof(status));
+}
+
+static void* event_thread(void* arg)
+{
+	int epoll_fd = *(int*)arg;
+
+	while (1) {
+		struct epoll_event event;
+		if (epoll_pwait(epoll_fd, &event, 1, -1, NULL) < 0)
+			break;
+
+		char buf[1024] = {0};
+		size_t buf_size = read(event.data.fd, buf, sizeof(buf));
+		if (buf_size > 0 && buf[0] == HCI_COMMAND_PKT)
+			process_command_pkt(event.data.fd, buf + 1, buf_size - 1);
+	}
+
+	return NULL;
+}
+
+#define ACL_HANDLE 256
+#define HCI_DEV 0
+
+static void initialize_vhci()
+{
+#if SYZ_EXECUTOR
+	if (!flag_vhci_injection)
+		return;
+#endif
+
+	int hci_socket = syz_init_net_socket(AF_BLUETOOTH, SOCK_RAW, BTPROTO_HCI);
+	if (hci_socket < 0)
+		fail("syz_init_net_socket failed");
+
+	vhci_fd = open("/dev/vhci", O_RDWR);
+	if (vhci_fd == -1)
+		fail("open /dev/vhci failed");
+
+	int epoll_fd = epoll_create1(0);
+	if (epoll_fd == -1)
+		fail("epoll_create1 failed");
+
+	struct epoll_event event = {0};
+	event.events = EPOLLIN;
+	event.data.fd = vhci_fd;
+	if (epoll_ctl(epoll_fd, EPOLL_CTL_ADD, vhci_fd, &event))
+		fail("epoll_ctl failed");
+
+	pthread_t th;
+	if (pthread_create(&th, NULL, event_thread, &epoll_fd))
+		fail("pthread_create failed");
+
+	// Register hci device
+	uint8 setup_cmd[2];
+	setup_cmd[0] = HCI_VENDOR_PKT;
+	setup_cmd[1] = 0;
+	if (write(vhci_fd, setup_cmd, sizeof(setup_cmd)) < 0)
+		fail("write failed");
+
+	// Bring hci device up
+	int res, tries = 16;
+	do {
+		res = ioctl(hci_socket, HCIDEVUP, HCI_DEV);
+		if (res != 0) {
+			if (errno == ENODEV)
+				continue;
+			else if (errno == EALREADY)
+				break;
+			else
+				fail("ioctl(HCIDEVUP) failed");
+		}
+	} while (res != 0 && tries--);
+
+	if (tries == 0)
+		fail("failed to bring hci device up");
+
+	// Activate page scanning mode which is required to fake a connection.
+	struct hci_dev_req dr = {0};
+	dr.dev_id = HCI_DEV;
+	dr.dev_opt = SCAN_PAGE;
+	if (ioctl(hci_socket, HCISETSCAN, &dr))
+		fail("ioctl(HCISETSCAN) failed");
+
+	// Fake a connection with bd address aa:aa:aa:aa:aa:aa.
+	// This is a fixed address used in sys/linux/socket_bluetooth.txt.
+	struct hci_ev_conn_request request;
+	memset(&request, 0, sizeof(request));
+	memset(&request.bdaddr, 0xaa, 6);
+	request.link_type = ACL_LINK;
+	hci_send_event_packet(vhci_fd, HCI_EV_CONN_REQUEST, &request, sizeof(request));
+
+	struct hci_ev_conn_complete complete;
+	memset(&complete, 0, sizeof(complete));
+	complete.status = 0;
+	complete.handle = ACL_HANDLE;
+	memset(&complete.bdaddr, 0xaa, 6);
+	complete.link_type = ACL_LINK;
+	complete.encr_mode = 0;
+	hci_send_event_packet(vhci_fd, HCI_EV_CONN_COMPLETE, &complete, sizeof(complete));
+
+	struct hci_ev_remote_features features;
+	memset(&features, 0, sizeof(features));
+	features.status = 0;
+	features.handle = ACL_HANDLE;
+	hci_send_event_packet(vhci_fd, HCI_EV_REMOTE_FEATURES, &features, sizeof(features));
+
+	close(epoll_fd);
```

Comment:
Closing file descriptors does not work this way. It does not necessary unblock other threads and worse there is simply no way this can work without introducing bad races (calling epoll_wait on a random fd), so no one tried to make this work reliably or at all.
You are assuming that the other thread is necessary inside of epoll_wait and is completely blocked. The other thread can be at an arbitrary point, including not inside of epoll_wait at all, in which case it will call epoll_wait later on a random fd potentially blocking forever.

Reply:
Thanks for catching this! The internet also suggests that that is a buggy behavior: https://stackoverflow.com/questions/26546858/what-happen-when-epoll-file-descriptor-is-closed